syntax = "proto3";

package aea.eightballer.hackathon_engagement.v1_0_0;

message HackathonEngagementMessage{

  // Custom Types
  message ErrorCode{
    enum ErrorCodeEnum {
      ERROR_CODE_ENUM_NOT_FOUND = 0;
      ERROR_CODE_ENUM_INVALID_SUBMISSION = 1;
      ERROR_CODE_ENUM_TEAM_FORMAT_ERROR = 2;
      ERROR_CODE_ENUM_INTERNAL_ERROR = 3;
    }
    ErrorCodeEnum error_code = 1;
  }

  message Status{
    enum StatusEnum {
      STATUS_ENUM_SUCCESS = 0;
      STATUS_ENUM_PENDING = 1;
      STATUS_ENUM_REJECTED = 2;
    }
    StatusEnum status = 1;
  }

  message SubmissionType{
    enum SubmissionTypeEnum {
      SUBMISSION_TYPE_ENUM_IDEA = 0;
      SUBMISSION_TYPE_ENUM_PROGRESS = 1;
      SUBMISSION_TYPE_ENUM_FINAL = 2;
    }
    SubmissionTypeEnum submission_type = 1;
  }


  // Performatives and contents
  message Idea_Proposal_Performative{
    string team_id = 1;
    string idea_description = 2;
  }

  message Team_Formation_Performative{
    string team_name = 1;
    repeated string member_invites = 2;
  }

  message Team_Formation_Response_Performative{
    string team_id = 1;
    Status status = 2;
  }

  message Update_Submission_Performative{
    string team_id = 1;
    string update_description = 2;
  }

  message Final_Submission_Performative{
    string team_id = 1;
    string project_link = 2;
    string documentation_link = 3;
    bool documentation_link_is_set = 4;
    string presentation_link = 5;
    bool presentation_link_is_set = 6;
  }

  message Submission_Acknowledgement_Performative{
    string team_id = 1;
    SubmissionType submission_type = 2;
    Status status = 3;
  }

  message Error_Performative{
    ErrorCode error_code = 1;
    string error_msg = 2;
  }


  oneof performative{
    Error_Performative error = 5;
    Final_Submission_Performative final_submission = 6;
    Idea_Proposal_Performative idea_proposal = 7;
    Submission_Acknowledgement_Performative submission_acknowledgement = 8;
    Team_Formation_Performative team_formation = 9;
    Team_Formation_Response_Performative team_formation_response = 10;
    Update_Submission_Performative update_submission = 11;
  }
}
